File: 2018/haskell/day01.hs
import qualified Data.IntSet as S

main :: IO ()
main = do
    changes <- parseInput <$> readFile "../data/input01.txt"
    print $ part1 changes
    print $ part2 changes

parseInput :: String -> [Int]
parseInput =  map (read . filter (/= '+')) . lines

part1 :: [Int] -> Int
part1 = sum

part2 :: [Int] -> Int
part2 = go S.empty . scanl (+) 0 . cycle
    where go s (x:xs)
              | x `S.member` s = x
              | otherwise = go (S.insert x s) xs

File: 2018/python/day01.py
#Advent of Code - Day 01

from itertools import cycle, accumulate

with open("../data/input01.txt") as f:

    changes = [int(n.strip()) for n in f.readlines()]
    print("Part 1: {}".format(sum(changes)))

    seen = set()
    print("Part 2: {}".format(next(f for f in accumulate(cycle(changes)) if f in seen or seen.add(f))))